### PR TITLE
7955: Replace PCSTor::CircuitIsolationKey with net::NetworkIsolationKey

### PR DESCRIPTION
**Pull request by mariospr**
Resolves https://github.com/brave/brave-browser/issues/7955

## Submitter Checklist:

- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue.
- [ ] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed).
- [ ] Requested a security/privacy review as needed.

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [x] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: net/proxy_resolution/proxy_config_service_tor.cc [hunk 1]
```diff
@@ -123,33 +124,12 @@ void ProxyConfigServiceTor::UpdateProxyURI(const std::string& uri) {
...
-  return domain;
-}
-
 void ProxyConfigServiceTor::SetNewTorCircuit(const GURL& url) {
   const HostPortPair old_host_port = proxy_server_.host_port_pair();
+
+  url::Origin url_origin = url::Origin::Create(url);
   const HostPortPair new_host_port(
-      CircuitIsolationKey(url),
+      net::NetworkIsolationKey(url_origin, url_origin).ToString(),
```
Comment (darkdh) [inline net/proxy_resolution/proxy_config_service_tor.cc:132]: ToString will give us "[origin] [origin]" , this is not what we what for username of socks5 config

Comment (mariospr) [reply net/proxy_resolution/proxy_config_service_tor.cc]: Yes, this was completely wrong and obviously the tests failed as well. Sorry for the sloppy first upload.

Just force-pushed the branch now and it's now passing all tests and working, PTAL again thanks!

**Comment by mariospr** [comment]
@bridiver Just added you as reviewer here as well as the person who originally requested this change. PTAL and let me know what you think, thanks!

File: net/proxy_resolution/proxy_config_service_tor.cc [hunk 2]
```diff
@@ -18,6 +18,7 @@
 #include "base/time/time.h"
 #include "base/values.h"
 #include "crypto/random.h"
+#include "net/base/network_isolation_key.h"
 #include "net/base/registry_controlled_domains/registry_controlled_domain.h"
```
Comment (darkdh) [inline net/proxy_resolution/proxy_config_service_tor.cc:22]: This header is no longer needed

Comment (mariospr) [reply net/proxy_resolution/proxy_config_service_tor.cc:22]: Fixed

File: chromium_src/net/base/network_isolation_key.h [hunk 3]
```diff
@@ -0,0 +1,20 @@
+/* Copyright (c) 2020 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#ifndef BRAVE_CHROMIUM_SRC_NET_BASE_NETWORK_ISOLATION_KEY_H_
+#define BRAVE_CHROMIUM_SRC_NET_BASE_NETWORK_ISOLATION_KEY_H_
+
+#define BRAVE_NETWORK_ISOLATION_KEY_H \
```
Comment (darkdh) [inline chromium_src/net/base/network_isolation_key.h:9]: can we do 
```
#define GetTopFrameOrigin             \
  GetEffectiveTopFrameOrigin() const; \
  base::Optional<url::Origin>& GetTopFrameOrigin
```
so that we don't need any patches into header and put the implementation in .cc override

Comment (mariospr) [reply chromium_src/net/base/network_isolation_key.h]: Good point, I do certainly prefer not to introduce a .patch file if not strictly needed. Fixed!

**Review by darkdh — APPROVED** [review]

**Comment by mariospr** [comment]
@bridiver All tests passed and @darkdh already approved but still needing a review from you, PTAL thanks!

File: net/proxy_resolution/proxy_config_service_tor.cc [hunk 4]
```diff
@@ -137,13 +137,13 @@ std::string ProxyConfigServiceTor::CircuitIsolationKey(const GURL& url) {
...
-      origin.host(),
-      net::registry_controlled_domains::INCLUDE_PRIVATE_REGISTRIES);
-  if (domain.size() == 0)
-    domain = origin.host();
-  return domain;
+  url::Origin url_origin = url::Origin::Create(url);
+  net::NetworkIsolationKey network_isolation_key(url_origin, url_origin);
+
+  const base::Optional<url::Origin>& network_isolation_key_origin =
+    network_isolation_key.GetEffectiveTopFrameOrigin();
```
Comment (bridiver) [inline net/proxy_resolution/proxy_config_service_tor.cc:144]: lint nit: 4 spaces for continuation of a line

File: chromium_src/net/base/network_isolation_key.cc [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+
+#include "../../../../net/base/network_isolation_key.cc"
+
+namespace net {
+
+const base::Optional<url::Origin>&
+NetworkIsolationKey::GetEffectiveTopFrameOrigin() const {
+  DCHECK_EQ(original_top_frame_origin_.has_value(),
+            top_frame_origin_.has_value());
+  return top_frame_origin_;
```
Comment (bridiver) [inline chromium_src/net/base/network_isolation_key.cc:16]: this seems extremely risky to me and I don't think we should make this change. Why wouldn't we just use the method as-is?

**Review by bridiver — CHANGES_REQUESTED** [review]
changing the return value of a method like GetTopFrameOrigin seems like a bad idea to me. This could have all kinds of unintended consequences and potentially break things or cause security issues

**Comment by mariospr** [comment]
@bridiver This CL is **not changing** the return value of `GetTopFrameOrigin()` at all. It's **adding** a new method called `GetEffectiveTopFrameOrigin()` (maybe not the best name, though) so that we can retrieve the internal `top_frame_origin_` value, which is the result of processing the original value passed to the `NetworkIsolationKey` constructor. 

This is necessary to implement the idea of using `net::NetworkIsolationKey` in `ProxyConfigServiceTor::CircuitIsolationKey` (which you suggested in https://github.com/brave/brave-core/pull/4374#discussion_r371976481) because `NetworkIsolationKey` only provides a getter for the **original** `top_frame_origin` passed to its constructor, but the one we need is the one that got processed upon construction via the `SwitchToRegistrableDomainAndRemovePort()` function, which called `GetDomainAndRegistry()` to strip out the subdomains from the origin.

Otherwise, if we just use the `GetTopFrameOrigin()` method that is currently available, we'll get several unit test failures:
```
...
../../brave/net/proxy_resolution/proxy_config_service_tor_unittest.cc:104: Failure
Expected equality of these values:
  expected_key
    Which is: "torproject.org"
  actual_key
    Which is: "check.torproject.org"
Stack trace:
#0 0x559700070502 net::ProxyConfigServiceTorTest_CircuitIsolationKey_Test::TestBody()
../../brave/net/proxy_resolution/proxy_config_service_tor_unittest.cc:104: Failure
Expected equality of these values:
  expected_key
    Which is: "bbc.co.uk"
  actual_key
    Which is: "www.bbc.co.uk"
Stack trace:
#0 0x559700070502 net::ProxyConfigServiceTorTest_CircuitIsolationKey_Test::TestBody()
```

@bridiver Please take another look into this when you have a moment and let me know if this clarifies things, thanks!

File: net/proxy_resolution/proxy_config_service_tor.cc (hunk 4 above)
Comment (mariospr) [reply net/proxy_resolution/proxy_config_service_tor.cc:144]: Oops! Sorry, will change that

File: chromium_src/net/base/network_isolation_key.cc (hunk 5 above)
Comment (mariospr) [reply chromium_src/net/base/network_isolation_key.cc:16]: Replied to this in the PR here: https://github.com/brave/brave-core/pull/7199#issuecomment-731553278

TL;DR: We need it because the currently available getter `GetTopFrameOrigin()` returns the original value passed to the constructor, unprocessed. And for the CircuitIsolationKey we need the value after having been processed with GetDomainAndRegistry(), which is different from the original value passed to the constructor.

**Review by bridiver — APPROVED** [review]

File: chromium_src/net/base/network_isolation_key.cc (hunk 5 above)
Comment (iefremov) [reply chromium_src/net/base/network_isolation_key.cc:16]: I was skimming the NIK sources and found `NetworkIsolationKey::CreateWithNewFrameOrigin`. Wondering, would it be possible to avoid override with a hack like this:
```
net::NetworkIsolationKey network_isolation_key(url_origin, url_origin);
net::NetworkIsolationKey real_nik = network_isolation_key.CreateWithNewFrameOrigin(url_origin);
```
looks like it does the trick?
@mariospr

Comment (mariospr) [reply chromium_src/net/base/network_isolation_key.cc:16]: @iefremov Good point! I did look into the very same function on Friday, but that doesn't work unfortunately in the code we have in master because the implementation of that method [still uses `top_frame_origin_` the create the second key](https://chromium.googlesource.com/chromium/src.git/+/refs/tags/87.0.4280.67/net/base/network_isolation_key.cc#93):
```
  NetworkIsolationKey NetworkIsolationKey::CreateWithNewFrameOrigin(
      const url::Origin& new_frame_origin) const {
    if (!top_frame_origin_)
      return NetworkIsolationKey();
    NetworkIsolationKey key(top_frame_origin_.value(), new_frame_origin);
    key.opaque_and_non_transient_ = opaque_and_non_transient_;
    return key;
  }
```
This said, I've just noticed that it seems that such `chromium_src` override won't be needed anymore once we rebase this PR on top of Chromium 88, because [`GetTopFrameSite()` does no longer return the original value in there](https://chromium.googlesource.com/chromium/src.git/+/refs/tags/88.0.4324.11/net/base/network_isolation_key.h#113), 
but [the one that has been processed right on the constructor](https://chromium.googlesource.com/chromium/src.git/+/refs/tags/88.0.4324.11/net/base/network_isolation_key.cc#235).

Thus, consider this chromium_src as a temporary thing, will go away from Cr88 on!

Comment (iefremov) [reply chromium_src/net/base/network_isolation_key.cc:16]: but isn't `top_frame_origin_` is what we actually want?.. (you return it in the newly added getter)
anyway, I'm fine with leaving it as is till cr88 - probably worth filing an issue?

Comment (mariospr) [reply chromium_src/net/base/network_isolation_key.cc:16]: Hmm.. I think you're right, sorry. I got confused with `original` and `origin` and thought it was returning `original_top_frame_origin_`. Now I'm not sure why this trick did not work when I tried it on Friday, maybe I did something else wrong and blamed it on this by mistake. My bad, sorry for the confusion.

As for filing an issue, just one comment on that this patch is only a thing in Cr87 and will go away once we rebase this fix on top of Cr88, where `GetTopFrameSite()` will return what we want already and we will be able to drop it. Still, I think filing an issue makes sense because we'll need to remember to drop this either way, or we'll risk leaving this override or the hack around for longer, so here it is: https://github.com/brave/brave-browser/issues/12869

Thanks, and apologies again for the confusion!
